Make invocation URL dynamic

skwashd opened the pull request:
*Issue #, if available:*

* #43 

*Description of changes:*

The API endpoint in Amazon's implementation of Lambda used the path `/2015-03-31/functions/[func]/invocations` to invoke the function. `[func]` the name of the function and is set as the value of the `AWS_LAMBDA_FUNCTION_NAME` environment variable when the function is executing on AWS.

The emulator uses a hard coded endpoint of `/2015-03-31/functions/function/invocations`. This is even the case
when the `AWS_LAMBDA_FUNCTION_NAME` environment variable is set to another value.

This patch changes the endpoint URL when the `AWS_LAMBDA_FUNCTION_NAME` environment variable is set. In this case the invocation URL will be `/2015-03-31/functions/${AWS_LAMBDA_FUNCTION_NAME}/invocations`. When the environment variable isn't set, the current behaviour persists and the function name is set to `function`.

The end result is the emulator behaviour is closer to the real environment the functions execute in.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

valerena reviewed (changes requested):
This change looks good, but there are conflicts in the test cases (we renamed the file and added testing capabilities for multiple architectures). If you bring those changes and fix the tests, we should be good to go.

skwashd commented:
@valerena thanks. Fixed, but before merging this you should review my comment on https://github.com/aws/aws-lambda-runtime-interface-emulator/pull/49#issuecomment-933664875.

skwashd commented:
@valerena bumping the thread as suggested in https://github.com/aws/aws-lambda-runtime-interface-emulator/pull/49#issuecomment-946872526

valerena commented:
Hi @skwashd , we had a discussion with the team and there are concerns about backwards compatibility and breaking existing customers that had automated local calls for testing using RIE. The consensus is that this behavior could be optional, having to pass an extra parameter to have it like this. 

To have this as the default behavior might come in the future, but it's something that we still have to confirm with more stakeholders and if it happens it could take several months, to make sure that we have all use cases covered and the right communication about that change. 

I know that this has already taken more time than what I imagined you originally expected, but if you change this Pull Request to work as an optional behavior I would work to prioritize its revision and include it in a shorter amount of time that what has taken to review the rest of the changes so far. 
Let me know if this sounds okay for you.

skwashd commented:
@valerena it's always "fun" trying to fix broken behaviour when people the (broken) behaviour 😢 

Before I spend more time on this, I'd like to make sure we're aligned on the details. 

Is the proposal to keep the default function name fix as implemented in ebc933d as is? I hope this is a less controversial change as the default function name should match the path in the url. I hope many users aren't depending on the `AWS_LAMBDA_FUNCTION_NAME` environment variable being set to `test_function`.

For the invoke URL change, I propose that if `AWS_LAMBDA_FUNCTION_NAME` is set, we have the server listen on both the hardcoded `function` path and the one generated from the environment variable. It will use the same handler function for both. If the variable isn't set, we simply listen on the `function` path. 

This approach keeps backwards compatibility and makes it easier to remove the broken behaviour down the road. I will ensure I minimise any duplicated logic. Does this work for you and the team? I'd like to see the old hardcoded `function` path url when the variable is set marked as deprecated in the docs, or is that too soon?

valerena commented:
@skwashd The team's stance on this is that RIE wasn't launched to be an exact copy of Lambda's Control Plane, so the current behavior is not "broken", and it's just "the way that RIE works". So, we shouldn't make any changes to the "default behavior", even if we think not many people use that. Every change that could break existing customers should be activated through a specific flag passed, and we won't change the recommended path url for now, even if the other url is also available. 

It's possible than during next year we will have a longer discussion about the future here and potentially add your recommendations as the default behavior, but at this moment we don't have the resources to start that conversation.

skwashd commented:
@valerena lets agree to disagree on inconsistent behaviour is a bug or a feature. I've incorporated the discussions we've discussed here. All the existing tests pass and I added tests to exercise the new functionality.

skwashd commented:
@valerena any update on this one?

swantzter commented:
Hi @valerena, just wanted to check in on this and see if there's any potential for progress, but given the commit history I assume the situation for your team hasn't gotten to a better state yet?

skwashd commented:
@valerena Can I please get an update on this?

antocapp commented:
Can you guys finally merge this please?

timini commented:
hit the button, merge it!

calh commented:
I'm getting into developing step functions, developing and testing locally along with executing local lambda functions using this.  This feature really should have been in the RIE from the first release.  I don't see how a hard coded `function` name is usable when doing test driven development.

Please merge this @valerena!  I really want to give AWS my money, but my wallet is stuck because it only responds to the name `Wallet`, not `function`.

skwashd commented:
Given the recent activity in the repo is appears the team has resources again. @valerena can I please get an update? It's been over 2 years since you last responded.

calh commented:
Hey @skwashd, would you be willing to pull AWS's commits into your `dynamic-url` branch and publish a release of your code?  You're the captain now!

skwashd commented:
@calh I'm not going to expend any more energy on this until there is a clear commitment from AWS to accept it. Given the radio silence from @valerena I don't think it is going to be accepted.

valerena commented:
Hi everyone. I apologize for not replying earlier and for the silence for so long. The last few months we've been able to work on this again, mostly on security-related improvements, but I'd say we are in a position today where we can actually review and try to bring this in.

I think last time I looked at this, I got stuck trying to suggest a better naming than `AWS_LAMBDA_RIE_INCONSISTENT_BEHAVIOUR`, because it's confusing and it's unclear what the inconsistency is. Also, I think a positive env var is more appropriate instead of having to set something to "false". I suggest changing it to `AWS_LAMBDA_RIE_CONSISTENT_FUNCTION_NAME`, that has to be set to `TRUE` (or to any value, really)

There are currently some conflicts (the tests have changed a lot), so if you can update the PR with those changes and the variable name, I can take a look and see how we can include this in an upcoming release (better late than never?).

Do you think you can do that, @skwashd ? And apologies again.

skwashd commented:
I went with `AWS_LAMBDA_RIE_DYNAMIC_FUNCTION_URL` as it is more descriptive and is a positive name. Calling it `AWS_LAMBDA_RIE_CONSISTENT_FUNCTION_NAME` is less clear. I look forward to it being merged soon. 

I'd love to see this fixed properly in v2. It shouldn't require two environment variables to make RIE handle URLs the same way as Lambda.

valerena commented on test/integration/testdata/main.py:
```diff
@@ -25,13 +25,13 @@ def check_env_var_handler(event, context):
 
 def assert_lambda_arn_in_context(event, context):
-    if context.invoked_function_arn == f"arn:aws:lambda:us-east-1:012345678912:function:{os.environ.get('AWS_LAMBDA_FUNCTION_NAME', 'test_function')}":
+    if context.invoked_function_arn == f"arn:aws:lambda:us-east-1:012345678912:function:{os.environ.get('AWS_LAMBDA_FUNCTION_NAME', 'function')}":
```
I think I found an issue with our tests thanks to this change. 
Because the integration tests are passing but here they should fail.

Your new test works correctly with this code, but there was already a previous test `test_lambda_function_arn_exists()` -> `arnexists-x86_64` that doesn't pass the new env var, so this ARN should have `test_function` instead (the old value).

Looking at the [output of the test run](https://github.com/aws/aws-lambda-runtime-interface-emulator/actions/runs/9861242832/job/27231798350?pr=46) `arnexists-x86_64` doesn't appear on the list of tests that were run, so it looks it might be ignoring the failing tests instead of fail overall.

I'll have to figure out this issue, but in the meantime you should change this to have different cases depending if the new env var was passed or not (probably just create a new handler here to be called form the new integ test).

The previous handler in this file also has the same issue, but because it's the branch that shouldn't happen, then it's not relevant and the test still succeed, but you should change that too.


Everything else looks good

valerena replied:
Digging a little deeper, it looks like this shouldn't really be a problem, because when running this we will always have a value for `AWS_LAMBDA_FUNCTION_NAME`, so in this code it will never go to the second parameter of the `os.environ.get()`, 

It's still weird that the previous test that uses this code doesn't show up on the output for the tests, so I'm still investigating what's going on there. 

Did you make this change because the tests failed if you didn't? Or did you change it before actually running the tests? (My point is that: either this change is making the old test fail, or this change shouldn't be needed).

valerena replied:
Of course it had to be a simple thing. The old test is  not running because the new test you added has the same name! (`def test_lambda_function_arn_exists`) so it's actually overwriting the old test, but it succeeds. 


What I would say is: don't change this file. This change doesn't actually affect the test because the env var is always set. We could probably just remove that second parameter in the test, but for now, it's better if you don't modify the file, so it doesn't get mixed up as a needed change.

valerena commented on test/integration/local_lambda/test_end_to_end.py:
```diff
@@ -257,5 +257,25 @@ def test_custom_client_context(self):
+            self.assertEqual(b'"My lambda ran succesfully"', r.content)
+
+
+    def test_lambda_function_arn_exists(self):
```
This one should be `test_lambda_function_arn_exists_consistent`

skwashd commented:
I'll try to look at this again in the coming days

retrry commented:
What is the status of this? This would make local development easier in some cases, where you use separate orchestration engine for starting lambdas. Now I have to override function name there when developing.
